prov/usnic/src/usdf_ep_dgram.c

```diff
@@ -335,6 +335,305 @@ static struct fi_ops_cm usdf_cm_dgram_ops = {
+		return -FI_ENODATA;
+	if (hints->ep_attr->max_order_war_size > defaults.max_order_war_size)
+		return -FI_ENODATA;
+	if (hints->ep_attr->max_order_waw_size > defaults.max_order_waw_size)
+		return -FI_ENODATA;
+
+out:
+	*fi->ep_attr = defaults;
+
+	return FI_SUCCESS;
+}
+
+int usdf_dgram_fill_dom_attr(struct fi_info *hints, struct fi_info *fi)
+{
+	struct fi_domain_attr defaults;
+
+	defaults = dgram_dflt_domain_attr;
+
+	if (!hints || !hints->domain_attr)
+		goto out;
+
+	switch (hints->domain_attr->threading) {
+	case FI_THREAD_UNSPEC:
+	case FI_THREAD_ENDPOINT:
+		break;
+	case FI_THREAD_FID:
+	case FI_THREAD_COMPLETION:
+	case FI_THREAD_DOMAIN:
+		defaults.threading = hints->domain_attr->threading;
+		break;
+	default:
+		return -FI_ENODATA;
+	}
+
+	switch (hints->domain_attr->control_progress) {
+	case FI_PROGRESS_UNSPEC:
+	case FI_PROGRESS_AUTO:
+		break;
+	case FI_PROGRESS_MANUAL:
+		defaults.control_progress =
+			hints->domain_attr->control_progress;
+		break;
+	default:
+		return -FI_ENODATA;
+	}
+
+	switch (hints->domain_attr->data_progress) {
+	case FI_PROGRESS_UNSPEC:
+	case FI_PROGRESS_MANUAL:
+		break;
+	default:
+		return -FI_ENODATA;
+	}
+
+	switch (hints->domain_attr->resource_mgmt) {
+	case FI_RM_UNSPEC:
+	case FI_RM_DISABLED:
+		break;
+	default:
+		return -FI_ENODATA;
+	}
+
+	switch (hints->domain_attr->mr_mode) {
+	case FI_MR_UNSPEC:
+	case FI_MR_BASIC:
+		break;
+	default:
+		return -FI_ENODATA;
+	}
+
+out:
+	*fi->domain_attr = defaults;
+
+	return FI_SUCCESS;
+}
+
+int usdf_dgram_fill_tx_attr(struct fi_info *hints, struct fi_info *fi,
+		struct usd_device_attrs *dap)
+{
+	struct fi_tx_attr defaults;
+	size_t entries;
+
+	defaults = dgram_dflt_tx_attr;
+
+	defaults.size = dap->uda_max_send_credits / defaults.iov_limit;
+
+	if (!hints || !hints->tx_attr)
+		goto out;
+
+	/* make sure we can support the capabilities that are requested */
+	if (hints->tx_attr->caps & ~USDF_DGRAM_CAPS)
+		return -FI_ENODATA;
+
+	/* clear the mode bits we don't support */
+	if (hints->tx_attr->mode)
+		defaults.mode &= hints->tx_attr->mode;
+
+	/* make sure the app supports our required mode bits */
+	if ((defaults.mode & USDF_DGRAM_REQ_MODE) != USDF_DGRAM_REQ_MODE)
+		return -FI_ENODATA;
+
+	defaults.op_flags |= hints->tx_attr->op_flags;
+
+	if ((hints->tx_attr->msg_order | USDF_DGRAM_MSG_ORDER) !=
+			USDF_DGRAM_MSG_ORDER)
+		return -FI_ENODATA;
+	if ((hints->tx_attr->comp_order | USDF_DGRAM_COMP_ORDER) !=
+			USDF_DGRAM_COMP_ORDER)
+		return -FI_ENODATA;
+
+	if (hints->tx_attr->inject_size > defaults.inject_size)
+		return -FI_ENODATA;
+
+	if (hints->tx_attr->iov_limit > defaults.iov_limit)
+		return -FI_ENODATA;
+	if (hints->tx_attr->size > dap->uda_max_send_credits)
+		return -FI_ENODATA;
+
+	/* make sure the values for iov_limit and size are within appropriate
+	 * bounds. if only one of the two was given, then set the other based
+	 * on:
+	 * 	max_credits = size * iov_limit;
```

Comment:
Separately, please check that we're handling the `iov_limit` correctly for the non-prefix case.  This calculation (and the setting of `defaults.iov_limit`) doesn't change based on whether `FI_MSG_PREFIX` is specified, so that means the only way we are doing it correctly is if the checks/logic in the prefix/non-prefix implementations of, e.g., `fi_send` are different.  We may want to update this logic instead of that logic.